src/coreclr/jit/hwintrinsiccodegenarm64.cpp

```diff
@@ -496,6 +496,29 @@ void CodeGen::genHWIntrinsic(GenTreeHWIntrinsic* node)
                 GetEmitter()->emitIns_R_R_R(ins, emitSize, targetReg, op1Reg, op2Reg, opt);
                 break;
 
+            case NI_AdvSimd_CompareEqual:
+            case NI_AdvSimd_Arm64_CompareEqual:
+            case NI_AdvSimd_Arm64_CompareEqualScalar:
+                if (intrin.op1->isContained())
+                {
+                    assert(HWIntrinsicInfo::SupportsContainment(intrin.id));
+                    assert(intrin.op1->IsVectorZero());
+
+                    GetEmitter()->emitIns_R_R(ins, emitSize, targetReg, op2Reg, opt);
```

Comment:
This isn't related to your change, but rather the ARM64 emitter more generally.

I find this confusing to read because `emitIns_R_R` implies its a "one operand" instruction with a register source and register destination. In practice, however, this is basically (for lack of a better name) `emitIns_R_R_Z`, which is `emitIns_R_R_R` where the second operand is effectively `REG_ZR`.

It would be nice, IMO, if this were handled "more gracefully", either via the register allocator assigned `REG_ZR` itself or by having some emitter method with a name that clearly shows the implicit `0` being used. However, that probably needs more input/thought.

CC. @echesakovMSFT, thoughts?

Reply:
As an alternative to introducing another API in the emitter (which I would argue against) I would specialize the instruction by introducing another opcode. Say, instead of `INS_cmeq` we could call it `INS_cmeq_zero` (we have the "pseudo"instructions already - e.g. `INS_ld1_4regs`).

Reply:
I think that makes it overall harder/more confusing to use.

I'd really expect we either have a specialized form that supports zero or a special `REG_ZR` and have `emitIns_R_R_R` handle it, as that is effectively what the operation is "actually doing".

Having `emitIns_R_R` is confusing because its still a two operand instruction and still takes `2 input registers`. The specialization is that when one of the inputs is `ZR`, we can encode it more optimally.

Reply:
Well, I think `REG_ZR` is even more confusing - given the implied instruction operand is a vector with zero valued lanes.

Let's keep it as `R_R` for now. And follow up after the change is merged if needed.